pkg/cmd/build/translator.go

```diff
@@ -0,0 +1,258 @@
+	}
+}
+
+func (rt registryTranslator) translate(line string) string {
+	if strings.Contains(line, constants.DevRegistry) {
+		result := rt.replacer.Replace(line, constants.DevRegistry, rt.userNs)
+		return result
+	}
+
+	if strings.Contains(line, constants.GlobalRegistry) {
+		result := rt.replacer.Replace(line, constants.GlobalRegistry, rt.globalNamespace)
+		return result
+	}
+	return line
+}
+
+type cacheMountTranslator struct {
+	projectHash        string
+	cacheMountRegex    *regexp.Regexp
+	hasIDRegex         *regexp.Regexp
+	targetExtractRegex *regexp.Regexp
+	mountRegex         *regexp.Regexp
+	targetRegex        *regexp.Regexp
+}
+
+func newCacheMountTranslator(repo, dockerfilePath, target string) cacheMountTranslator {
+	cacheMountRegex := regexp.MustCompile(`^RUN.*--mount=.*type=cache`)
+	hasIDRegex := regexp.MustCompile(`^RUN.*--mount=[^ ]*id=`)
+	targetExtractRegex := regexp.MustCompile(`--mount=[^ ]*target=([^, ]+)`)
+	mountRegex := regexp.MustCompile(`--mount=([^[:space:]]+)`)
+	targetRegex := regexp.MustCompile(`target=([^,\s]+)`)
+
+	return cacheMountTranslator{
+		projectHash:        generateProjectHash(repo, dockerfilePath, target),
+		cacheMountRegex:    cacheMountRegex,
+		hasIDRegex:         hasIDRegex,
+		targetExtractRegex: targetExtractRegex,
+		mountRegex:         mountRegex,
+		targetRegex:        targetRegex,
+	}
+}
+
+func generateProjectHash(repositoryURL, manifestName, path string) string {
+	// Create input string for hashing
+	input := fmt.Sprintf("%s-%s-%s", repositoryURL, manifestName, path)
+
+	// Generate SHA256 hash
+	hasher := sha256.New()
+	hasher.Write([]byte(input))
+	hash := hasher.Sum(nil)
+
+	// Return first 12 characters of hex hash for readability
+	return hex.EncodeToString(hash)[:12]
+}
+
+func (cmt cacheMountTranslator) translate(line string) string {
+
+	// Check if this RUN command has a cache mount
+	if !cmt.cacheMountRegex.MatchString(line) {
+		return line
+	}
+
+	// If an id is already defined, leave unchanged
+	if cmt.hasIDRegex.MatchString(line) {
+		return line
+	}
+
+	// Find all --mount= occurrences and process them individually
+	result := line
+
+	// Find all mount definitions
+	matches := cmt.mountRegex.FindAllStringSubmatch(result, -1)
+	for _, match := range matches {
+
+		if len(match) < 2 {
+			continue
+		}
+
+		fullMount := match[0]   // --mount=type=cache,target=...
+		mountParams := match[1] // type=cache,target=...
+
+		// Check if this is a cache mount
+		if !strings.Contains(mountParams, "type=cache") {
+			continue
+		}
+
+		// Check if it already has an id
+		if strings.Contains(mountParams, "id=") {
+			continue
+		}
+
+		// Extract target from this specific mount
+		target := ""
+		if targetMatch := cmt.targetRegex.FindStringSubmatch(mountParams); len(targetMatch) > 1 {
+			target = targetMatch[1]
+		}
+
+		// Generate unique ID for this mount
+		id := cmt.projectHash
+		if target != "" {
+			id = fmt.Sprintf("%s-%s", cmt.projectHash, target)
```

Comment:
Shouldn't the `target` being part of the hash as it happens with test caches? (maybe I'm wrong, but I'd swear that in the test case, this was part of the hash instead of being independent